avail-server/src/main/java/com/avail/server/io/WebSocketChannel.kt

```diff
@@ -56,15 +59,54 @@ import java.nio.channels.AsynchronousSocketChannel
  *   The [WebSocketAdapter].
  * @param transport
  *   The [channel][AsynchronousSocketChannel].
+ * @param heartbeatInterval
+ *   The time in milliseconds between each heartbeat request made by the server
+ *   to the client.
+ * @param heartbeatTimeout
+ *   The amount of time, in milliseconds, after which the heartbeat will fail if
+ *   a heartbeat is not received from the client by the server.
+ * @param heartbeatFailureThreshold
+ *   The number of consecutive times the `heartbeatTimeout` is allowed to be
+ *   reached before disconnecting the client.
+ * @param onChannelCloseAction
+ *   The custom action that is to be called when the input channel is closed in
+ *   order to support implementation-specific requirements during the closing of
+ *   a channel.
  */
 internal class WebSocketChannel constructor(
 	override val adapter: WebSocketAdapter,
-	override val transport: AsynchronousSocketChannel)
+	override val transport: AsynchronousSocketChannel,
+	heartbeatFailureThreshold: Int,
+	heartbeatInterval: Long,
+	heartbeatTimeout: Long,
+	onChannelCloseAction:
+		(DisconnectReason, AbstractTransportChannel<AsynchronousSocketChannel>)
+			-> Unit)
 : AbstractTransportChannel<AsynchronousSocketChannel>()
 {
 	override val isOpen get() = transport.isOpen
 	override val maximumSendQueueDepth = MAX_QUEUE_DEPTH
 	override val maximumReceiveQueueDepth = MAX_QUEUE_DEPTH
+	override val heartbeat: Heartbeat =
+		WebSocketChannelHeartbeat(
+			this,
+			heartbeatFailureThreshold,
+			heartbeatInterval,
+			heartbeatTimeout)
+
+	/**
+	 * `true` indicates the close [onChannelCloseAction] has not been run and
+	 * is eligible to run; `false` otherwise.
+	 */
+	private val onChannelCloseActionNotRun = AtomicBoolean(true)
+
+	override val onChannelCloseAction: (DisconnectReason) -> Unit =
+		{
+			if (!onChannelCloseActionNotRun.getAndSet(false))
+			{
+				onChannelCloseAction(it, this)
+			}
```

Comment:
The condition here is reversed. You only want to run `onChannelCloseAction` once, the very first time. `getAndSet` answers the current value of the `AtomicBoolean` and then sets the new value. Since you initialized `onChannelCloseActionNotRun` to `true`, this will cause the predicate to fail the first time and pass every subsequent time, repeatedly calling `onChannelCloseAction` only when it shouldn't.

Reply:
Yup, this is due to a renaming I did to actually remove the `!`. I renamed the field and its direction but forgot to remove the !